File: Difficulty: Medium/Gas Station/gas-station.cpp
class Solution {
  public:
    int startStation(vector<int> &gas, vector<int> &cost) {
        //  code here
        int n = gas.size();  
  
        // Variables to track total and current remaining gas
        int totalGas = 0;
        int currGas = 0;
        int startIdx = 0;
    
        // Traverse through each station to calculate remaining
        // gas in the tank, and total gas 
        for(int i = 0; i < n; i++) {
          	currGas += gas[i] - cost[i]; 
            totalGas += gas[i] - cost[i];  
    
            // If currGas is negative, circular tour can't 
          	// start with this index, so update it to next one
            if(currGas < 0) {
                currGas = 0;
                startIdx = i + 1; 
            }
        }
    
        // No solution exist
        if(totalGas < 0) 
            return -1;
    
        return startIdx;
    }
};

File: Difficulty: Medium/Gas Station/gas-station.java
class Solution {
    public int startStation(int[] gas, int[] cost) {
        // code here
        int n = gas.length;
        int startIdx = 0;

        // Initially tank is empty
        int currGas = 0;

        for (int i = 0; i < n; i++) {
            currGas = currGas + gas[i] - cost[i];

            // If currGas becomes less than zero, then
            // It's not possible to proceed with this startIdx
            if (currGas < 0) {
                startIdx = i + 1;
                currGas = 0;
            }
        }

        // Checking if startIdx can be a valid 
    	// starting point for the Circular tour
        currGas = 0;
        for (int i = 0; i < n; i++) {

            // Circular Index
            int idx = (i + startIdx) % n;
            currGas = currGas + gas[idx] - cost[idx];
            if (currGas < 0)
                return -1;
        }

        return startIdx;
    }
}